app/scripts/metamask-controller.js

```diff
@@ -1783,14 +1785,21 @@ export default class MetamaskController extends EventEmitter {
     // Ensure preferences + identities controller know about all addresses
     this.preferencesController.addAddresses(addresses)
     this.accountTracker.syncWithAddresses(addresses)
+  }
 
-    const wasLocked = !isUnlocked
-    if (wasLocked) {
-      const oldSelectedAddress = this.preferencesController.getSelectedAddress()
```

Comment:
I'm not sure about this stuff being moved. The `wasLocked` variable implies that this only ran in the case where the keyring controller _was_ locked but is no longer, but... I don't understand how that could be the case. The `state` passed in is the _current_ state, and it's checking that it's _not_ unlocked, so... this doesn't seem equivalent?

Reply:
>>As part of some permissions-related functionality, I need to reliably emit an event every time the extension becomes unlocked

>This should be possible.

>> I understand that `addresses.length === 0 if` the extension/keyring controller is locked (I understand those to be one and the same), but it’s unclear to me why `addresses.length > 0 && !isUnlocked` would ever be `true`

>Looking at the KC code, only only time an 'update' event is emitted with isUnlocked set to false is after a call to #setLocked, which also unsets/empties the keyrings. So, yes, `addresses.length > 0 && !isUnlocked` would never be true, with the current state of the code. (https://github.com/MetaMask/KeyringController/blob/5a95355fa4c0856c0f8d7c0cb57fa6af91364482/index.js#L114-L126)

>BUT

>No, that's wrong. All calls to updateState would also emit 'update' events, so lines 121 and 124 will each emit an `'update'` event (https://github.com/MetaMask/KeyringController/blob/5a95355fa4c0856c0f8d7c0cb57fa6af91364482/index.js#L121-L124) the first with `isUnlocked === false && keyrings.length > 0`, the second with `isUnlocked === false && keyrings.length === 0`

Reply:
All of which is to say, I think it is equivalent?

Reply:
I don't quite follow. That makes me thing that this block was executed _on lock_, not on _unlock_.

Reply:
Hm. On review, I agree, but I _think_ the attention was the opposite?
- https://github.com/MetaMask/metamask-extension/pull/5066
- The commit that added the `wasLocked` block: https://github.com/MetaMask/metamask-extension/pull/5066/commits/003d445a98164dac0c0529dfd69f5e3987d7d05c
- The commit that added the tests: https://github.com/MetaMask/metamask-extension/pull/5066/commits/e803b8e047b5acad9143fe28d99d9e7d65211f46

If the extension is locked, we'll see `keyrings === []`, and there will be no accounts to select. I interpreted this code as trying to cover for the case where a keyring had been removed since the `wasLocked` (now `_onUnlock`) block was executed.

That can still happen with a `keyringController` state update, I suppose, but I don't see how it would happen while the keyring controller/extension is locked.

I'm seeing if e2e tests pass if the block is deleted.

Reply:
Yeah I think you're right. I don't understand why this would be on-lock either.
@whymarrh do you recall?

Reply:
Actually it looks like this is covered already by the `this.preferencesController.syncAddresses(accounts)`call in `submitPassword`. That ensures the accounts are valid upon unlock. I don't think we need this stuff after all.

Reply:
Resolved via #8224